app/src/main/java/woowacourse/shopping/shopping/ProductsAdapter.kt

```diff
@@ -0,0 +1,51 @@
+package woowacourse.shopping.shopping
+
+import android.view.ViewGroup
+import androidx.recyclerview.widget.RecyclerView
+import woowacourse.shopping.model.ProductUIModel
+import woowacourse.shopping.shopping.viewHolder.ItemViewHolder
+import woowacourse.shopping.shopping.viewHolder.ProductsViewHolder
+import woowacourse.shopping.shopping.viewHolder.ReadMoreViewHolder
+import woowacourse.shopping.shopping.viewHolder.RecentProductsViewHolder
+
+class ProductsAdapter(
+    productItemTypes: List<ProductsItemType>,
+    private val onClickItem: (data: ProductUIModel) -> Unit,
+    private val onReadMoreClick: () -> Unit
+) : RecyclerView.Adapter<ItemViewHolder>() {
+    private var productItemTypes: MutableList<ProductsItemType> = productItemTypes.toMutableList()
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemViewHolder {
+        return when (viewType) {
+            ProductsItemType.TYPE_HEADER -> RecentProductsViewHolder.from(parent, onClickItem)
+            ProductsItemType.TYPE_FOOTER -> ReadMoreViewHolder.from(parent) { onReadMoreClick() }
+            ProductsItemType.TYPE_ITEM -> ProductsViewHolder.from(parent) {
+                onClickItem((productItemTypes[it] as ProductItem).product)
+            }
+            else -> throw IllegalArgumentException("Invalid view type")
+        }
+    }
+
+    override fun getItemCount(): Int = productItemTypes.size
+
+    override fun onBindViewHolder(holder: ItemViewHolder, position: Int) {
+        when (holder) {
+            is RecentProductsViewHolder -> holder.bind(productItemTypes[position])
+            is ProductsViewHolder -> holder.bind(productItemTypes[position])
+            is ReadMoreViewHolder -> return
+        }
+    }
+
+    override fun getItemViewType(position: Int): Int {
+        return when (productItemTypes[position]) {
+            is RecentProductsItem -> ProductsItemType.TYPE_HEADER
+            is ProductItem -> ProductsItemType.TYPE_ITEM
+            is ProductReadMore -> ProductsItemType.TYPE_FOOTER
+        }
+    }
+
+    fun updateData(data: List<ProductsItemType>) {
+        productItemTypes = data.toMutableList()
+        notifyDataSetChanged()
```

Comment:
IDE에서 다음과 같은 경고문을 보셨을 거예요! 더 나은 방법이 있을까요?
> It will always be more efficient to use more specific change events if you can. Rely on notifyDataSetChanged as a last resort.

힌트: 현재는 전체 데이터셋을 업데이트할 필요가 없습니다. 새로 추가되는 데이터에 대해서만 업데이트 요청을 보내면 됩니다.

Reply:
onBinding에 로그를 찍어보면서 어떤 차이점이 있는지 학습 해보았습니다!!